engine_redis.go

```diff
@@ -1854,19 +1854,40 @@ const (
 	leavePushType pushType = 2
 )
 
-func extractPushData(data []byte) ([]byte, pushType, uint64) {
+var (
+	metaSep    = []byte("__")
+	contentSep = ":"
+)
+
+// See tests for supported format examples.
+func extractPushData(data []byte) ([]byte, pushType, StreamPosition) {
 	var offset uint64
-	if bytes.HasPrefix(data, []byte("__")) {
-		parts := bytes.SplitN(data, []byte("__"), 3)
-		if bytes.Equal(parts[1], []byte("j")) {
-			return parts[2], joinPushType, 0
-		} else if bytes.Equal(parts[1], []byte("l")) {
-			return parts[2], leavePushType, 0
+	var epoch string
+	if bytes.HasPrefix(data, metaSep) {
+		pos := bytes.Index(data[len(metaSep):], metaSep)
+		if pos > 0 {
+			content := data[len(metaSep) : len(metaSep)+pos]
+			rest := data[len(metaSep)+pos+len(metaSep):]
+			if bytes.Equal(content, []byte("j")) {
+				return rest, joinPushType, StreamPosition{}
+			} else if bytes.Equal(content, []byte("l")) {
+				return rest, leavePushType, StreamPosition{}
+			}
```

Comment:
those 2 `if` can be rewrittent with a `switch` or just without an `else`

Reply:
fixed, also refactored this function – now handling cases with invalid data